workspaces/github-discussions/plugins/search-backend-module-github-discussions/src/collators/GithubDiscussionsCollatorFactory.ts

```diff
@@ -0,0 +1,245 @@
+      'clearCacheOnSuccess',
+    );
+    const discussionsBatchSize = _config.getOptionalNumber(
+      'discussionsBatchSize',
+    );
+    const commentsBatchSize = _config.getOptionalNumber('commentsBatchSize');
+    const repliesBatchSize = _config.getOptionalNumber('repliesBatchSize');
+
+    return new GithubDiscussionsCollatorFactory({
+      logger,
+      credentialsProvider,
+      githubIntegration,
+      timeout,
+      url,
+      cacheBase,
+      clearCacheOnSuccess,
+      discussionsBatchSize,
+      commentsBatchSize,
+      repliesBatchSize,
+    });
+  }
+
+  async getCollator() {
+    return Readable.from(this.execute());
+  }
+
+  async *execute(): AsyncGenerator<GithubDiscussionsDocument> {
+    const logger: typeof console = {
+      log: this.logger.info.bind(this.logger),
+      info: this.logger.info.bind(this.logger),
+      error: this.logger.error.bind(this.logger),
+      warn: this.logger.warn.bind(this.logger),
+      dir: this.logger.info.bind(this.logger),
+      debug: this.logger.debug.bind(this.logger),
+    } as unknown as typeof console;
+
+    const url = gh(this.url);
+    assert(url !== null, `Not parsable as a Github URL`);
+    const { name: repo, owner: org } = url;
+    assert(org !== null, `Discussions url is missing organization name`);
+    assert(repo !== null, `Discussion url is missing repository`);
+
+    const integration = this.githubIntegration.byUrl(url.href);
+
+    assert(integration, `Could not retrieve a Github integration for ${url}`);
+
+    const { token } = await this.credentialsProvider.getCredentials({
+      url: url.href,
+    });
+
+    const client = createGithubGraphqlClient({
+      token,
+      endpoint: `${integration?.config.apiBaseUrl}/graphql`,
+    });
+
+    let documents: AsyncIterable<GithubDiscussionFetcherResult> | undefined;
+    const cache = this.cacheBase ? new URL(this.cacheBase) : undefined;
+    const {
+      timeout,
+      clearCacheOnSuccess,
+      discussionsBatchSize = 100,
+      commentsBatchSize = 100,
+      repliesBatchSize = 100,
+    } = this;
+
+    const task = run(function* injection(): Operation<void> {
+      const scope = yield* useScope();
+      const results = createQueue<GithubDiscussionFetcherResult, void>();
+
+      yield* spawn(function* (): Operation<void> {
+        try {
+          yield* fetchGithubDiscussions({
+            client,
+            org,
+            repo,
+            discussionsBatchSize,
+            commentsBatchSize,
+            repliesBatchSize,
+            logger,
+            results,
+            cache,
+            timeout,
+            clearCacheOnSuccess,
+          });
+        } catch (e) {
+          logger.log(e);
+          logger.error(
+            `Encountered an error while ingesting GitHub Discussions`,
+            e,
+          );
+        }
+        results.close();
+      });
+
+      documents = toAsyncIterable(results, scope);
+
+      yield* suspend();
```

Comment:
this isn't a pattern we use in this repository or in the core Backstage. Is it possible to simplifying this using normal promises or native async iterators? Tbh it looks complex and hard to follow